### ISSUE TITLE
hide boot messages when plymouth is enabled

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#32556]
## Issue description

Plymouth is used to prevent boot messages from being displayed. That does not work completely.

/cc @abbradar @michaelpj

### Steps to reproduce

NixOS system with Grub 2 and Xfce.

1. `boot.plymouth.enable = true;`
1. `nixos-rebuild switch`
1. `reboot`
3. you still get these messages

![virtualbox_nixos_11_12_2017_06_02_11](https://user-images.githubusercontent.com/91113/33817108-f664e8e8-de3d-11e7-8259-eed1a4c5500e.png)

## Technical details
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.66, NixOS, 17.09.2281.b4a0c011e81 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.15`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by gnidorah** [comment]
Yep, not great. With latest nixos-unstable
On boot: grub -> few stage 1 messages -> Plymouth splash -> few stage 1 messages for a moment -> tty login prompt for a moment -> display manager
On reboot: blinking cursor for some time -> Plymouth splash for a moment

**Comment by michaelpj** [comment]
Yeah, currently we install Plymouth into the stage 2 of the initrd. So you don't get it until then, and if if stage 2 is very fast, you won't see much of it at all! I don't know why it goes into stage 2, @abbradar set it up initially, so might know.

**Comment by gnidorah** [comment]
Also `boot.plymouth.theme.default` background is black so black text "NixOS" on the `boot.plymouth.logo.default` is invisible.
Better to change `boot.plymouth.theme.default` background color to one used in `services.xserver.displayManager.slim.theme.default`.

**Comment by gnidorah** [comment]
And to fix
> tty login prompt for a moment -> display manager

We need to provide systemd units for each display manager. See https://wiki.archlinux.org/index.php/plymouth#Smooth_transition

**Comment by rileyinman** [comment]
This is still important, it just has not been worked on.

**Comment by davidak** [comment]
This is still needed for good UX.

**Comment by peterhoeg** [comment]
I have patches here, but cleaned up yet and ready for submission. With the latest plymouth we can show the UEFI vendor logo during boot which is quite nifty. Also just sitting in my local patch queue.......

**Comment by bobvanderlinden** [comment]
@peterhoeg that sounds great! Did you submit your patches somewhere already?

**Comment by peterhoeg** [comment]
Crucially I had left out a "not" in my earlier comment that should have been "but *not* cleaned up yet and ready for submisssion"... So there's that.

I'll allocate some time for this Soonish(tm). I'll ping you guys here when it's ready for testing. I have it working on one laptop here but not another although I'm not sure if this is related to what I did or the laptop in question.

**Comment by maydayv7** [comment]
(just bumping the issue)

**Comment by daanhenke** [comment]
Does anyone know a fix for this issue on 22.11? Or is it impossible to install the package into the stage1 initramfs?

**Comment by maydayv7** [comment]
I'm using `systemd-boot` with GDM and plymouth with the following configuration on 21.11:

```nix
{
  boot = {
    consoleLogLevel = 0;
    initrd.verbose = false;
    plymouth.enable = true;
    kernelParams = [
      "quiet"
      "splash"
      "boot.shell_on_fail"
      "i915.fastboot=1"
      "loglevel=3"
      "rd.systemd.show_status=false"
      "rd.udev.log_level=3"
      "udev.log_priority=3"
    ];

    loader = {
      timeout = lib.mkDefault 0;
      efi.canTouchEfiVariables = true;
      systemd-boot = {
        enable = true;
        editor = false;
        configurationLimit = 100;
      };
    };
  };
}

```

It works flawlessly, displays the logo, without any text on boot/shutdown

**Comment by TLATER** [comment]
> It works flawlessly, displays the logo, without any text on boot/shutdown

That "flawlessly" somewhat arguable, the point of plymouth in my mind has always been that you *can* get logs by pressing escape, in case you need them, so disabling all logging seems a bit counter that :)

Using [boot.initrd.systemd.enable](https://search.nixos.org/options?channel=22.05&show=boot.initrd.systemd.enable&from=0&size=50&sort=relevance&type=packages&query=boot.initrd.systemd) is probably the way forward,  but from my experimentation this still shows the stage1 logs (and a line stating something along the lines of "starting plymouth"), followed by some console flickering when switching to lightdm.

**Comment by pmarreck** [comment]
Stuck on grub2 for now due to zfs-on-root (although if someone wants to provide a guide to get zfs-on-root working with systemd-boot, that'd be great!), and would also like the option to see a nicer boot and making console text optional. When you work on your nixos config for hours to get things just right, this kind of thing is the icing on the cake.

I don't use LUKS, so I'm fine with an option that doesn't yet work with LUKS (some protection in place for those who do but try to enable plymouth, might be warranted)

**Comment by pmarreck** [comment]
EDIT: I got it to work. Plymouth is now uninterrupted. My (current) config is on my profile, I'm not sure exactly which combination of bits makes it work but one of the configs suggested in the recent comments helped. Note that a previous incarnation/attempt at this actually hid a real error/problem (regarding mounting my bpool in time), and the only symptom of that was that it kept dumping me into a root login at boot which I could just control-D out of, so be aware that you might want to turn it off sometimes.

**Comment by mainrs** [comment]
> [...] so be aware that you might want to turn it off sometimes.

So pressing escape to see the boot logs is not possible?

**Comment by pmarreck** [comment]
@mainrs 
I don't think that worked for me unfortunately. I too would like that feature or something like it (like holding down Shift, etc.)

**Comment by pmarreck** [comment]
@mainrs 
Nevermind, escape DID work! It does show the full boot text after hitting it. So everything is sweet now :)

**Comment by heywoodlh** [comment]
Just FYI, this is my config to suppress boot messages on my machines:

```
  boot = {
    kernelParams = [ "quiet" "splash" ];
    plymouth.enable = true;
    consoleLogLevel = 0;
    initrd.verbose = false;
  };
```

**Comment by davidak** [comment]
@heywoodlh that has no effect on my system using grub2. Do you use systemd-boot?

https://www.youtube.com/watch?v=4_C8l3YhGwM

**Comment by heywoodlh** [comment]
@davidak Ah yes, sorry, should have clarified: I am using systemd-boot.

**Comment by TLATER** [comment]
> https://www.youtube.com/watch?v=4_C8l3YhGwM

Most of that appears to be grub showing stuff, which happens before even the Linux kernel takes over, and won't be affected by any of the systemd/plymouth configuration. That's up to configuring grub.

After grub finishes "booting the kernel" you get a blank screen for a while, which I assume is most of stage 1. It's blank because you've disabled all logging for that portion of the boot - ideally plymouth would start much earlier here, so you don't need to disable stage 1 logging (and can make it visible through escape).

Finally just before stage 1 hands over to stage 2 plymouth finally starts (so that it can ask for decryption stuff), but as lightdm takes over you get tty1 flickering again. I'm pretty sure it's possible to resolve that as well, I've seen fedora systems boot without that flicker.

I think the current state is:

- [x] You can make the system boot without showing the boot manager
  - To achieve this use systemd-boot
  - OR configure grub to be hidden by default:
    ```nix
    {
      boot.loader = {
        timeout = 5; # The default, I recommend making this shorter
        grub = {
          extraConfig = ''
            set timeout_style=hidden
          '';
          # Because it still draws that image otherwise
          splashImage = null;
        };
      };
    }
    ```
- [x] You can, with caveats, make the system fully boot with plymouth from stage1
  - [x] Completely feasible without full disk encryption, just set `boot.plymouth.enable = true;`
  - [ ] With full disk encryption, currently requires the experimental `boot.initrd.systemd`
- [ ] You can kind of hack the boot such that it shows no boot logs during stage1 to get a console-log free boot
  - See [above](https://github.com/NixOS/nixpkgs/issues/32556#issuecomment-1378261367)
  - This is definitely not optimal
- [ ] You cannot make the system boot without flickering the tty for a bit before your display manager starts

**Comment by codeiimon** [comment]
I was having a similar issue and just wanted to post my solution to this. I'm currently using grub on an intel system and could only get plymouth to show on system shutdown but not on system boot. What fixed it for me was add the 'i915' kernel module to my initrd.kernalModules like so:

`boot.initrd.kernelModules = [ "i915" ];`

What led me to the answer was this reddit post linked [here](https://www.reddit.com/r/archlinux/comments/i7ag9a/comment/g10sduf/).

Also, I'm not sure if this is necessary, but I also added these lines of code to hide logs on boot:

```
boot.consoleLogLevel = 0;
boot.initrd.verbose = false;
boot.initrd.kernelParams = [
        "quiet"
        "splash"
        "boot.shell_on_fail"
        "i915.fastboot=1"
        "loglevel=3"
        "rd.systemd.show_status=false"
        "rd.udev.log_level=3"
        "udev.log_priority=3"
];
```

**Comment by jmickelin** [comment]
@codeiimon Thank you! This comment led me down the right track. It also does work without the lines to disable the logs, for what it's worth. `i915.fastboot=1` seems to prevent the display of the console again between Plymouth closing and the display manager starting.

An alternative method that doesn't require adding any new modules in the initrd image is to force Plymouth to use the SimpleDRM driver. By default it hesitates to do so (waiting for some udev probing to detect "better" options), but reluctantly ends up using it after an annoyingly long timeout. You can tell it to go for SimpleDRM immediately by adding `plymouth.use-simpledrm` to the kernel command line:

```
boot.kernelParams = [
  ...
  plymouth.use-simpledrm
  ...
];
```

(It seems that the Plymouth developers are considering making this the default in future versions: https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/110)

Here is a full, minimal, example:
```
  boot.loader.systemd-boot.enable = true; # For reference, since I see no reason why this wouldn't work with other bootloaders.

  boot.initrd.systemd.enable = true; # (optional, experimental) Needed to show the decryption prompt in Plymouth.
  boot.plymouth.enable = true;
  boot.kernelParams = [
    "plymouth.use-simpledrm"
  ];
```

(note that `"splash"` isn't strictly needed either, since it should already be added by `boot.plymouth.enable = true;`).

---

As for the flicker free boot sequence mentioned in previous posts, I found this summary of how Fedora does it on Intel GPU machines, though it might be out of date compared to what they're currently doing: https://fedoraproject.org/wiki/Changes/FlickerFreeBoot It relies on that `i915.fastboot` parameter. I assume this means that the initrd does need to bundle the `i915` module.

**Comment by musjj** [comment]
@jmickelin 

I added:

```nix
kernelParams = [
  "plymouth.use-simpledrm"
];
```

But it still shows boot messages between the first splash and the login manager.

**Comment by musjj** [comment]
I just took this module from Misterio77's config: https://github.com/Misterio77/nix-config/blob/a8727347957891bd651aba3bc0e04ca694d6f664/hosts/common/optional/quietboot.nix and it's working flawlessly for me.

I'd encourage anyone who failed to extract the right incantations from the rune fragments left in this thread to try it out.

**Comment by mainrs** [comment]
> This is something that always happens even with the default config (without plymouth: replace "plymouth" above with "boot logs"). But I wonder if there's a way to eliminate this black screen gap.

I'm shooting in the dark here, but there is a feature called smooth transition. I only know of GDM that supports this feature.

**Comment by musjj** [comment]
> but there is a feature called smooth transition.

Thanks, I looked it up and it seems interesting. I'm already using GNOME though.

For whatever reason, it no longer shows a frozen Plymouth frame after the monitor blackout transition, and instead just a flat black (lit) screen, which is a huge improvement. I really don't know what caused this though and I don't think I did anything with my config.

Also, during shutdown, there's a very *tiny* chance that an extremely brief frame of a 1-2 line console log will be shown. I couldn't read it fast enough though. But Plymouth works *most* of the time.

**Comment by musjj** [comment]
I've managed to capture and transcribe the logs I see on shutdown:

```bash
[  257.826441] (sd-umoun[5052]: Failed to unmount /run/shutdown/mounts/38f870ec48758433: Device or resource busy
[  257.828854] watchdog: watchdog0: watchdog did not stop!
[  257.828909] shutdown[1]: Unable to finalize remaining filesystems, ignoring.
```

I've tried all options above but I can never get rid of this log message. I can't even find these lines in `journalctl` for some reason, had to record and transcribe it by hand.

Anyone have any clue how to hide, get rid of or fix these messages?
